src/viewport/index.tsx

```diff
@@ -596,12 +595,24 @@ class Viewport extends React.Component<
         }
     }
 
-    public changeAgentsColor(colorChange: ColorChange): void {
-        const { agent, color } = colorChange;
-        const agentIds = this.selectionInterface.getAgentIdsByNamesAndTags([
-            agent,
-        ]);
-        this.visGeometry.applyColorToAgents(agentIds, color);
+    public changeAgentsColor(appliedColors: UIDisplayData): void {
+        appliedColors.forEach((agent) => {
+            const agentIds = this.selectionInterface.getAgentIdsByNamesAndTags([
+                    { name: agent.name, tags: [] },
+                ]);
+            this.visGeometry.applyColorToAgents(
+                agentIds,
+                agent.color
+            );
+
+            agent.displayStates.forEach((state) => {
+                const stateIds =
+                    this.selectionInterface.getAgentIdsByNamesAndTags([
+                        { name: agent.name, tags: [state.name] },
+                    ]);
+                this.visGeometry.applyColorToAgents(stateIds, state.color);
+            });
```

Comment:
Under the hood, `applyColorToAgents` is calling `VisGeometry.updateScene`, which looks like it may be expensive. Now that this call is happening in a loop, is there a way to ensure that the update just happens once, at the end of this function?

Reply:
Yes and why not, its cleaner! Also even if `updateScene` is relatively cheap (cheap enough to get called in every animate loop at least), it might get more expensive in some upcoming work on caching, so I like the idea of minimizing calls where we can.